# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/generic/nsIFrame.cpp
# Commit: 6d76b71d89f3
# Full Hash: 6d76b71d89f3d3218632b68cecf9c7295b4fb41b
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2024-04-04 03:44:04
# Regressor Bug: 1887954
# File Overlap Count: 2
# Description:
#   Bug 1887954 - Make `nsIFrame::GetLastLeaf()` never returns a frame in native anonymous subtrees under the given frame r=emilio
#   
#   It does not check whether the child frame is in a native anonymous subtree
#   only when every first child at digging the frame tree.  This patch rewrites
#   the complicated loop with 2 for-loops and make it always check whether the
# ==============================================================================

diff -r 3187c4f491a8 -r 6d76b71d89f3 layout/generic/nsIFrame.cpp
--- a/layout/generic/nsIFrame.cpp	Wed Apr 03 07:15:55 2024 +0000
+++ b/layout/generic/nsIFrame.cpp	Wed Apr 03 07:20:57 2024 +0000
@@ -10716,21 +10716,27 @@
 }
 
 void nsIFrame::GetLastLeaf(nsIFrame** aFrame) {
-  if (!aFrame || !*aFrame) return;
-  nsIFrame* child = *aFrame;
-  // if we are a block frame then go for the last line of 'this'
-  while (1) {
-    child = child->PrincipalChildList().FirstChild();
-    if (!child) return;  // nothing to do
-    nsIFrame* siblingFrame;
-    nsIContent* content;
-    // ignore anonymous elements, e.g. mozTableAdd* mozTableRemove*
-    // see bug 278197 comment #12 #13 for details
-    while ((siblingFrame = child->GetNextSibling()) &&
-           (content = siblingFrame->GetContent()) &&
-           !content->IsRootOfNativeAnonymousSubtree())
-      child = siblingFrame;
-    *aFrame = child;
+  if (!aFrame || !*aFrame) {
+    return;
+  }
+  for (nsIFrame* maybeLastLeaf = (*aFrame)->PrincipalChildList().LastChild();
+       maybeLastLeaf;) {
+    nsIFrame* lastChildNotInSubTree = nullptr;
+    for (nsIFrame* child = maybeLastLeaf; child;
+         child = child->GetPrevSibling()) {
+      nsIContent* content = child->GetContent();
+      // ignore anonymous elements, e.g. mozTableAdd* mozTableRemove*
+      // see bug 278197 comment #12 #13 for details
+      if (content && !content->IsRootOfNativeAnonymousSubtree()) {
+        lastChildNotInSubTree = child;
+        break;
+      }
+    }
+    if (!lastChildNotInSubTree) {
+      return;
+    }
+    *aFrame = lastChildNotInSubTree;
+    maybeLastLeaf = lastChildNotInSubTree->PrincipalChildList().LastChild();
   }
 }
 